library/kani/src/models/mod.rs

```diff
@@ -0,0 +1,192 @@
+            impl MaskElement for $ty {
+                const TRUE: Self = -1;
+                const FALSE: Self = 0;
+            }
+        }
+    }
+
+    macro_rules! impl_unsigned_element {
+        { $ty:ty } => {
+            impl MaskElement for $ty {
+                const TRUE: Self = <$ty>::MAX;
+                const FALSE: Self = 0;
+            }
+        }
+    }
+
+    impl_element! { i8 }
+    impl_element! { i16 }
+    impl_element! { i32 }
+    impl_element! { i64 }
+    impl_element! { i128 }
+    impl_element! { isize }
+
+    impl_unsigned_element! { u8 }
+    impl_unsigned_element! { u16 }
+    impl_unsigned_element! { u32 }
+    impl_unsigned_element! { u64 }
+    impl_unsigned_element! { u128 }
+    impl_unsigned_element! { usize }
+
+    pub(super) const fn mask_len(len: usize) -> usize {
+        (len + 7) / 8
+    }
+
+    #[cfg(target_endian = "little")]
+    unsafe fn simd_bitmask_impl<T, const LANES: usize>(input: &[T; LANES]) -> [u8; mask_len(LANES)]
+    where
+        T: MaskElement,
+    {
+        let mut mask_array = [0; mask_len(LANES)];
+        for lane in (0..input.len()).rev() {
+            let byte = lane / 8;
+            let mask = &mut mask_array[byte];
+            let shift_mask = *mask << 1;
+            *mask = if input[lane] == T::TRUE {
+                shift_mask | 0x1
+            } else {
+                assert_eq!(input[lane], T::FALSE, "Masks values should either be 0 or 1");
+                shift_mask
+            };
+        }
+        mask_array
+    }
+
+    /// Stub for simd_bitmask.
+    ///
+    /// It will reduce a simd vector (TxN), into an integer of size S (in bits), where S >= N.
+    /// Each bit of the output will represent a lane from the input. A lane value of all 0's will be
+    /// translated to 1b0, while all 1's will be translated to 1b1.
+    ///
+    /// In order to be able to do this pragmatically, we take additional parameters that are filled
+    /// by the compiler.
+    #[rustc_diagnostic_item = "KaniModelSimdBitmask"]
+    pub(super) unsafe fn simd_bitmask<T, U, E, const LANES: usize>(input: T) -> U
+    where
+        [u8; mask_len(LANES)]: Sized,
+        E: MaskElement,
+    {
+        // These checks are compiler sanity checks to ensure we are not doing anything invalid.
+        assert_eq!(
+            size_of::<U>(),
+            size_of::<[u8; mask_len(LANES)]>(),
+            "Expected size of return type and mask lanes to match",
+        );
+        assert_eq!(
+            size_of::<T>(),
+            size_of::<Simd::<E, LANES>>(),
+            "Expected size of input and lanes to match",
+        );
+
+        let data = &*(&input as *const T as *const [E; LANES]);
+        let mask = simd_bitmask_impl(data);
+        (&mask as *const [u8; mask_len(LANES)] as *const U).read()
+    }
+
+    /// Structure used for sanity check our parameters.
+    #[repr(simd)]
+    struct Simd<T, const LANES: usize>([T; LANES]);
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
Can you add a description to each test? It's difficult to see what they're doing.

Reply:
Sure